Document Antrea versioning (#1352)
vmware-tanzu/antrea: docs/versioning.md

patch support. This may seem short, but was done on purpose to encourage users
to upgrade Antrea often and avoid potential incompatibility issues. In the
future, we may reduce our release cadence for minor releases and simultaneously
increase the support windows for each release.

## Antrea upgrade and supported version skew

Our goal is to support "graceful" upgrades for Antrea. By "graceful", we notably
mean that there should be no significant disruption to data-plane connectivity
nor to policy enforcement, beyond the necessary disruption incurred by the
restart of individual components:

 * during the Antrea Controller restart, new policies will not be
   processed. Because the Controller also runs the validation webhook for
   [Antrea-native policies](antrea-network-policy.md), an attempt to create an
   Antrea-native policy resource before the restart is complete may return an
   error.
 * during an Antrea Agent restart, the Node's data-plane will be impacted: new
   connections to & from the Node will not be possible, and existing connections
   may break.

In particular, it should be possible to upgrade Antrea without compromising

> enforcement of existing network policies for both new and existing Pods.

Author: Don't we use `NetworkPolicies` specifically for the K8s resource name? How do we usually refer to security policies as a whole (K8s & Antrea-native)?

> enforcement of existing network policies for both new and existing Pods.

Reviewer: makes sense

upgraded to one of 0.11, 0.12, 0.13 or 0.14, but we discourage direct upgrades
to 0.15 and beyond. With the current release cadence, this provides a 6-month
window of compatibility. If we reduce our release cadence in the future, we may
revisit this policy as well.

When directly applying a newer Antrea YAML manifest, as provided for each
[release](https://github.com/vmware-tanzu/antrea/releases), there is no
guarantee that the Antrea Controller will be upgraded first. In practice, the
Controller would be upgraded simultaneously with the first Agents to be upgraded
by the rolling update of the Agent DaemonSet. This may create some transient
issues and compromise the "graceful" upgrade. For upgrade scenarios, we
therefore recommend that you "split-up" the manifest to ensure that the
Controller is upgraded first.

## Supported K8s versions

Each Antrea minor release should support [maintained K8s
releases](https://kubernetes.io/docs/setup/release/version-skew-policy/#supported-versions)
at the time of release (3 up to K8s 1.19, 4 after that). For example, at the
time that Antrea 0.10 was released, the latest K8s version was 1.19; as a result

> we guarantee that 0.10 supports at least 1.19, 1.18 and 1.17 (in practice is

Reviewer: ```suggestion
we guarantee that 0.10 supports at least 1.19, 1.18 and 1.17 (in practice it
```

also supports K8s 1.16).

In addition, we strive to support the K8s versions used by default in

> cloud-managed K8s services ([EKS], [AKS] and [GKE (regular channel)]). 

Reviewer 2: [GKE (regular channel)] doesn't appear as link

## Deprecation policies

### Prometheus metrics deprecation policy

Antrea follows a similar policy as

> [Kubernetes](https://cloud.google.com/kubernetes-engine/docs/release-notes) for

Reviewer 3: I could not find any Prometheus deprecation policy for K8s in the link https://cloud.google.com/kubernetes-engine/docs/release-notes. Wondering if that is the intended link.

Author: Thanks. Looks like I got my links confused, will fix

metrics deprecation.

Alpha metrics have no stability guarantees; as such they can be modified or
deleted at any time.

Stable metrics are guaranteed to not change; specifically, stability means:

 * the metric itself will not be renamed
 * the type of metric will not be modified

Eventually, even a stable metric can be deleted. In this case, the metric must
be marked as deprecated first and the metric must stay deprecated for at least

> one minor release. The [CHANGELOG] must announced both metric deprecations and

Reviewer: ```suggestion
one minor release. The [CHANGELOG] must announce both metric deprecations and
```

some_counter 0
```

In the future, we may introduce the same concept of [hidden
metric](https://kubernetes.io/docs/concepts/cluster-administration/system-metrics/#show-hidden-metrics)
as K8s, as an additional part of the metric lifecyle.

### APIs deprecation policy

The Antrea APIs are built using K8s (they are a combination of
CustomResourceDefinitions and aggregation layer APIServices) and we follow the
same versioning scheme as the K8s APIs and the same [deprecation
policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).

Other than the most recent API versions in each track, older API versions must
be supported after their announced deprecation for a duration of no less than:

 * GA: 12 months
 * Beta: 9 months
 * Alpha: N/A (can be removed immediately)

This also applies to the `controlplane` API. In particular, introduction and
removal of new versions for this API must respect the ["graceful" upgrade
guarantee](#antrea-upgrade-and-supported-version-skew). The `controlplane` API
(which is exposed using the aggregation layer) is often referred to as an
"internal" API as it is used by the Antrea components to communicate with each

> other, and is usually noit consummed by end users, e.g. cluster admins. However,

Reviewer: ```suggestion
other, and is usually not consumed by end users, e.g. cluster admins. However,
```

this API may also be used for integration with other software, which is why we
abide to the same deprecation policy as for other more "user-facing" APIs
(e.g. Antrea-native policy CRDs).

K8s has a [moratorium](https://github.com/kubernetes/kubernetes/issues/52185) on
the removal of API object versions that have been persisted to storage. At the

> moment, node of Antrea APIServices (which use the aggregation layer) persist

Reviewer 2: Do you mean "none of Antrea APIServices"?
